### PR TITLE
added liquibase changes: create tables roles, users, users_roles; add…

### PR DESCRIPTION
**Pull request by YuraKhilko**
…ed models: User, Role; added UserRepository, UserService (with implementation); added RegistrationException; added AuthenticationController with register and login endpoints + corresponding DTO with fields validation; added JWT validation.

### CODE REVIEW

File: src/main/java/bookstore/controller/BookController.java [hunk 1]
```diff
@@ -24,33 +25,37 @@
...
 @RequiredArgsConstructor
 @RestController
-@RequestMapping(value = "/api/books")
+@RequestMapping(value = "/books")
 public class BookController {
     private final BookService bookService;
 
     @Operation(summary = "Get all books", description = "Get a list of all available books")
     @GetMapping
+    @PreAuthorize("hasRole('USER')")
```
Comment (Rommelua) [inline src/main/java/bookstore/controller/BookController.java:34]: do you think ADMIN should not be allowed to use this end point?

File: src/main/java/bookstore/dto/user/UserLoginRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package bookstore.dto.user;
+
+import bookstore.validation.Email;
+import jakarta.validation.constraints.NotBlank;
+
+public record UserLoginRequestDto(
```
Comment (Rommelua) [inline src/main/java/bookstore/dto/user/UserLoginRequestDto.java]: why it is not a class?

File: src/main/java/bookstore/model/User.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return true;
```
Comment (Rommelua) [inline src/main/java/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/bookstore/repository/user/UserRepository.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package bookstore.repository.user;
+
+import bookstore.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface UserRepository extends JpaRepository<User, Long> {
+    @Query("FROM User u LEFT JOIN FETCH u.roles WHERE u.email = :email")
```
Comment (Rommelua) [inline src/main/java/bookstore/repository/user/UserRepository.java]: - **LEFT** is useless

- AND u.isDeleted = false AND r.isDeleted = false

File: src/main/java/bookstore/validation/Email.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import jakarta.validation.Payload;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Constraint(validatedBy = EmailValidator.class)
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.PARAMETER, ElementType.FIELD})
+public @interface Email {
```
Comment (Rommelua) [inline src/main/java/bookstore/validation/Email.java]: Why do you need custom? We have existed jakarta.validation.constraints.Email

File: src/main/java/bookstore/validation/FieldMatchValidator.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        Object fieldValue = new BeanWrapperImpl(value)
+                .getPropertyValue(field);
+        Object fieldMatchValue = new BeanWrapperImpl(value)
+                .getPropertyValue(fieldMatch);
+
+        if (fieldValue != null) {
+            return fieldValue.equals(fieldMatchValue);
+        } else {
+            return fieldMatchValue == null;
+        }
```
Comment (Rommelua) [inline src/main/java/bookstore/validation/FieldMatchValidator.java]: Objacts.equals?

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/bookstore/controller/BookController.java (hunk 1 above)
Comment (YuraKhilko) [reply src/main/java/bookstore/controller/BookController.java:34]: We have N:N connection User:Role. So, if user with Admin role wants to get info from books, role User should be ordered. Furthermore after discussing, made a decision to give a default role User when while a new User creating.

File: src/main/java/bookstore/dto/user/UserLoginRequestDto.java (hunk 2 above)
Comment (YuraKhilko) [reply src/main/java/bookstore/dto/user/UserLoginRequestDto.java]: made UserLoginResponseDto and UserLoginRequestDto as classes.

File: src/main/java/bookstore/validation/Email.java (hunk 5 above)
Comment (YuraKhilko) [reply src/main/java/bookstore/validation/Email.java]: Didn't know that. Thank you!

**Pull request merged**